gremlin-go/driver/resultSet.go

```diff
@@ -50,25 +50,60 @@ type channelResultSet struct {
 	statusAttributes map[string]interface{}
 	closed           bool
 	err              error
-	mux              sync.Mutex
+	waitSignal       chan bool
+	channelMux       sync.Mutex
+	waitSignalMux    sync.Mutex
+}
+
+func (channelResultSet *channelResultSet) sendSignal() {
+	// Lock wait
+	channelResultSet.waitSignalMux.Lock()
+	if channelResultSet.waitSignal != nil {
+		channelResultSet.waitSignal <- true
+		channelResultSet.waitSignal = nil
+	}
+	channelResultSet.waitSignalMux.Unlock()
 }
 
 func (channelResultSet *channelResultSet) GetError() error {
 	return channelResultSet.err
 }
 
 func (channelResultSet *channelResultSet) IsEmpty() bool {
-	channelResultSet.mux.Lock()
-	defer channelResultSet.mux.Unlock()
-	return channelResultSet.closed && len(channelResultSet.channel) == 0
+	channelResultSet.channelMux.Lock()
+	// If our channel is empty and we have no data in it, wait for signal that the state has been updated.
+	channelIsEmpty := len(channelResultSet.channel) == 0
+	if !channelIsEmpty || channelResultSet.closed {
+		// Channel is not empty.
+		channelResultSet.channelMux.Unlock()
+		return channelIsEmpty && channelResultSet.closed
+	} else {
+		// Channel is empty and not closed. Need to wait for signal that state has changed, otherwise
+		// we do not know if it is empty or not.
+		// We need ot grab the wait signal mutex before we release the channel mutex.
+		channelResultSet.waitSignalMux.Lock()
+		channelResultSet.channelMux.Unlock()
+
+		// Create a wait signal and unlock the wait signal mutex.
+		waitSignal := make(chan bool)
+		channelResultSet.waitSignal = waitSignal
+		channelResultSet.waitSignalMux.Unlock()
+
+		// Technically if we assigned channelResultSet.waitSignal then unlocked, it could be set to nil or
+		// overwritten to another channel before we check it, so to be safe, create additional variable and
+		// check that instead.
+		<-waitSignal
```

Comment:
What is the purpose of checking this?

Reply:
This is the async await flow that golang provides. Simply put, this thread creates a channel and grabs the data in it. The other thread writes to the channel. That means that this thread will be stuck here waiting until the other thread puts data into it, basically creating a wait block.